### ISSUE TITLE
does not find some things on Subdivx

### ISSUE THREAD

**Issue by rmichelena** [morpheus65535/bazarr#2378]
Bazarr 1.4.1

I have configured many providers, among them OS (.com) and Subdivx

manual search for Inspector Montalbano S09E03 only returns results from OS

<img width="1345" alt="Screenshot 2024-02-06 at 14 27 12" src="https://github.com/morpheus65535/bazarr/assets/77797875/73042fa9-5cf8-4c76-9de4-15f1133f1e69">

however checking out Subdivx, there are Spanish subs there too (and are better!)

<img width="990" alt="Screenshot 2024-02-06 at 14 29 13" src="https://github.com/morpheus65535/bazarr/assets/77797875/2339aaa7-bc6c-47d2-bed6-aac6cae658bc">

things to think about:
- Subdivx does not code language, everything there is supposed to be only Spanish, actually Latin American Spanish
- on Subdivx the subs are under "Il Commissario Montalbano", which is one of the alternate titles that Sonarr shows (and the actual original title), so that should be ok
- I also tried another show "Candice Renoir", this one does not have alternate titles, is the same everywhere, and for this one it does find subtitles on subdivx
-  other Montalbano episodes find subs on other providers too (addic7ed, gestdown) but never subdivx

so could it be that on Subdivx it's only searching the main (English) title, and not the alternate ones? (among which is the original title too!)

**Comment by morpheus65535** [comment]
I can confirm that subdivx provider isn't using alternative_titles at all while searching, only the main title.

This would be a nice improvement but requires to rework the search logic. I'm open to PR if you're willing to give it a try. I don't have time for this right now.

**Comment by rmichelena** [comment]
is this only on Subdivx, or are many providers searching only by main title? main title (e.g. English if your Sonarr is configured so) is so often not the original one, which is the one most likely to get results...

**Comment by morpheus65535** [comment]
> is this only on Subdivx, or are many providers searching only by main title? main title (e.g. English if your Sonarr is configured so) is so often not the original one, which is the one most likely to get results...

Short answer: it's up to each provider to decide.

Long answer: The way subliminal (on top of which Bazarr is built) let each provider implement requests to websites. This is done using the "query" method that may use video.title alone or video.alternative_titles if required. Many providers search using both but some don't.

**Comment by rmichelena** [comment]
I understand Bazarr grabs the "main title" and "alternative titles" from Sonarr. Does Bazarr pass all this to Subliminal?
I've examined the "opensubtitlescom.py" file inside "providers" and don't see any reference to alternative titles, iow it seems to make a single search, and I guess it's OS the one that has it's own list of alternative titles...
(I was trying to infer differences between "opensubtitlescom.py" and "subdivx.py", and it seems the problem is not there...)
so basically the question is, if I wanted to help, which files should I be looking at...?

**Comment by rmichelena** [comment]
there's a Video class that seems to be the object passed from Bazarr to Subliminal, and it'd seem that class does not implement any alternative_titles property...
so it'd seem absolutely no providers are searched with Alternative Titles, unless the provider (website) itself does it's own internal multi-title search, as seemingly Opensubtitles does ?

**Comment by rmichelena** [comment]
ok... I'm cheating, using ChatGPT + "AskTheCode" plugin to analyze... it says that when Bazarr constructs the Video object, the refine_from_db "refiner" actually adds the alternative_titles to the Video object... so that only the subdivx.py "provider" script would have to be intervened, to make multiple searches iterating over the multiple titles, and then return all the results as one batch - am I correct in this ?

**Comment by rmichelena** [comment]
sorry to be using this thread as a discovery log... so indeed I found out that other provider scripts, such as podnapisi.py, addic7ed.py and subscene.py, do use the video.alternative_titles and video.alternative_series in their queries ... so I'm trying to modify subdivx.py to do the same :) (opensubtitlescom.py was my first choice, and it happens that it does not use those properties, I guess it relies on the fact that opensubtitles does its own internal title matching).

**Comment by rmichelena** [comment]
ok, noob question... for testing out my modifications, I just replace the "subdivx.py" file inside the /app/bin/libs/subliminal_patch/providers/ directory inside the bazarr docker ? or do I have to restart the container?

**Comment by rmichelena** [comment]
Ok I've made some progress... I've been able to modify the Subdivx provider to use also the alternative titles... but it seems I've broken the logic that discards the non-matching season-episode combinations... so, some more work to do :)

**Comment by rmichelena** [comment]
DAMN IT!! just when I had nailed it, Subdivx totally revamped their site. Now nothing works, not my version nor the previous one. I don't think I'll be able to fix it, one thing is to have good logic and some python, another is to do webscraping scripts.

**Comment by morpheus65535** [comment]
It changed that much?

**Comment by rmichelena** [comment]
Totally... but now it should be simpler: the existing script was using GET with a query-constructed URL, and receiving HTML and parsing/scraping from it, and handling multiple pages.
Now it's a POST request with query payload, and receives a response in JSON with all subtitles at once.
So it should be much easier, but it's some work. I'm on it. ChatGPT is the one that's been helping me from the start, but lately it's become kinda reluctant to just write the complete code, nowadays it goes more like "here's an example" or "here's the concept".

**Comment by morpheus65535** [comment]
Keep up the good work! Thanks!

**Comment by rmichelena** [comment]
Ok !! code could be improved, but seems to work fine - finds and downloads.
I have no idea how to post it to the repository, so I've just attached it here.
[subdivx_v5.py.zip](https://github.com/morpheus65535/bazarr/files/14235767/subdivx_v5.py.zip)

**Comment by rmichelena** [comment]
[subdivx_v6.py.zip](https://github.com/morpheus65535/bazarr/files/14270550/subdivx_v6.py.zip)

they made a change again, subtle, but here taken into account

**Comment by morpheus65535** [comment]
Should be good in upcoming beta. Thanks!

**Comment by vitiko98** [comment]
@rmichelena A lot of my checks/heuristcs code I've been working on for years for multiple use cases was discarded by you. And you did all to make a single use case work. Incredible. You broke all the tests too.
